drivers/flash/nrf_qspi_nor.c

```diff
@@ -1121,7 +1185,7 @@ static int qspi_nor_write(const struct device *dev, off_t addr,
 			qspi_wait_for_completion(dev, res);
 		}
 	}
-	qspi_unlock(dev);
+	qspi_release(dev);
 
 	int res2 = qspi_nor_write_protection_set(dev, true);
```

Comment:
I think there is more work needed, by example of these two lines:
 - qspi_release(dev) at line 1188 puts device into DPD, as a part of release.
 - then qspi_nor_write_protection_set(dev, true) will invoke qspi_send_cmd, that will qspi_acquire (and do DPD wake up) and then the qspi_release (and DPD sleep again);

This is a lot of stuff getting sent to put into sleep and wake up the devices just between these two op.

Reply:
This is a situation I expected, and had opted for a simpler implementation, at the expense of possible performance.

I'm worried that the implementation will become more error prone if we split the mutex and the DPD handling, that it could be relatively easy to unintentionally leave the device in DPD.

A potential option would be to have separate "wake" and "sleep" calls, called at the beginning and end of things, similar to the ANOMALY_122 handling, and have a similar ref counter, when the counter hits zero, put the device back to sleep. Would that handle the issues you're concerned about here?

Reply:
That is a lot of extra comm on SPI lines, taking over QSPI peripheral and rescheduling via k_sem_unlock.
This is very aggressive power management at expense of performance.

Reply:
Updated to use PM_DEVICE_RUNTIME